java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java

```diff
@@ -17,81 +17,242 @@
-  }
+    private boolean allocated = false;
+
+    /**
+     * Note: fields are not closed when the writer is closed
+     *
+     * @param schema
+     * @param out
+     * @param allocator
+     */
+    protected ArrowWriter(Schema schema, OutputStream out, BufferAllocator allocator) {
+        this(schema.getFields(), createVectors(schema.getFields(), allocator), Channels.newChannel(out), true);
+    }
+
+    protected ArrowWriter(Schema schema, WritableByteChannel out, BufferAllocator allocator) {
+        this(schema.getFields(), createVectors(schema.getFields(), allocator), out, true);
+    }
+
+    protected ArrowWriter(List<Field> fields, List<FieldVector> vectors, OutputStream out) {
+        this(fields, vectors, Channels.newChannel(out), false);
+    }
+
+    protected ArrowWriter(List<Field> fields, List<FieldVector> vectors, WritableByteChannel out, boolean allocated) {
+        this.vectors = vectors;
+        this.out = new WriteChannel(out);
+        this.allocated = allocated;
+
+        // translate dictionary fields from in-memory format to message format
+        // add dictionary ids, change field types to dictionary type instead of index type
+        List<Field> updatedFields = new ArrayList<>(fields);
+        List<ArrowDictionaryBatch> dictionaryBatches = new ArrayList<>();
+        Set<Long> dictionaryIds = new HashSet<>();
+
+        // go through to add dictionary id to the schema fields and to unload the dictionary batches
+        for (FieldVector vector: vectors) {
+            if (vector instanceof DictionaryVector) {
+                Dictionary dictionary = ((DictionaryVector) vector).getDictionary();
+                long dictionaryId = dictionary.getId();
+                Field field = vector.getField();
+                // find the dictionary field in the schema
+                Field schemaField = null;
+                int fieldIndex = 0;
+                while (fieldIndex < fields.size()) {
+                    Field toCheck = fields.get(fieldIndex);
+                    if (field.getName().equals(toCheck.getName())) { // TODO more robust comparison?
+                        schemaField = toCheck;
+                        break;
+                    }
+                    fieldIndex++;
+                }
+                if (schemaField == null) {
+                    throw new IllegalArgumentException("Dictionary field " + field + " not found in schema " + fields);
+                }
+
+                // update the schema field with the dictionary type and the dictionary id for the message format
+                ArrowType dictionaryType = dictionary.getVector().getField().getType();
+                Field replacement = new Field(field.getName(), field.isNullable(), dictionaryType, dictionary.getEncoding(), field.getChildren());
+
+                updatedFields.remove(fieldIndex);
+                updatedFields.add(fieldIndex, replacement);
```

Comment:
@elahrvivaz I'd suggest that we decouple the Dictionary from the DictionaryVector for now. For context, one of the goals of dictionary encoding is to work with dictionary ids directly (to do aggregations, joins, etc faster). We don't always need the Dictionary around when doing so. 
The DictionaryVector can have a reference to the dictionary id instead of the dictionary itself. You could have a DictionaryProvider { Dictionary getDictionary(ID) } that returns the dictionary for the the id. That would remove the dependency on having a reference to  the Dictionary to create the DictionaryVector and still allow decoding a vector by looking up the corresponding dictionary.